test/validation-error-handling.test.js

```diff
@@ -336,3 +336,244 @@ test('should return a defined output message parsing JOI error details', t => {
+      hello: 'michelangelo'
+    },
+    url: '/'
+  }, (err, res) => {
+    t.error(err)
+    t.deepEqual(res.json(), {
+      statusCode: 400,
+      error: 'Bad Request',
+      message: 'my error'
+    })
+    t.strictEqual(res.statusCode, 400)
+  })
+})
+
+test('should catch error inside formatter and return message', t => {
+  t.plan(3)
+
+  const fastify = Fastify({
+    schemaErrorFormatter: (errors, dataVar) => {
+      throw new Error('abc')
+    }
+  })
+
+  fastify.post('/', { schema }, function (req, reply) {
+    reply.code(200).send(req.body.name)
+  })
+
+  fastify.inject({
+    method: 'POST',
+    payload: {
+      hello: 'michelangelo'
+    },
+    url: '/'
+  }, (err, res) => {
+    t.error(err)
+    t.deepEqual(res.json(), {
+      statusCode: 400,
+      error: 'Bad Request',
+      message: 'schema custom formatter error: abc'
+    })
+    t.strictEqual(res.statusCode, 400)
+    t.end()
+  })
+})
+
+test('cannot create a fastify instance with wrong type of errorFormatter', t => {
+  t.plan(2)
+
+  try {
+    Fastify({
+      schemaErrorFormatter: async (errors, dataVar) => {
+        return 'should not execute'
+      }
+    })
+  } catch (err) {
+    t.equals(err.message, 'schemaErrorFormatter option should not be an async function')
+  }
+
+  try {
+    Fastify({
+      schemaErrorFormatter: 500
+    })
+  } catch (err) {
+    t.equals(err.message, 'schemaErrorFormatter option should be a function, instead got number')
+  }
+})
+
+test('should register a route based schema error formatter', t => {
+  t.plan(3)
+
+  const fastify = Fastify()
+
+  fastify.post('/', {
+    schema,
+    schemaErrorFormatter: (errors, dataVar) => {
+      throw new Error('abc')
+    }
+  }, function (req, reply) {
+    reply.code(200).send(req.body.name)
+  })
+
+  fastify.inject({
+    method: 'POST',
+    payload: {
+      hello: 'michelangelo'
+    },
+    url: '/'
+  }, (err, res) => {
+    t.error(err)
+    t.deepEqual(res.json(), {
+      statusCode: 400,
+      error: 'Bad Request',
+      message: 'schema custom formatter error: abc'
+    })
+    t.strictEqual(res.statusCode, 400)
+    t.end()
+  })
+})
+
+test('prefer route based error formatter over global one', t => {
+  t.plan(3)
+
+  const fastify = Fastify({
+    schemaErrorFormatter: (errors, dataVar) => {
+      throw new Error('abc')
+    }
+  })
+
+  fastify.post('/', {
+    schema,
+    schemaErrorFormatter: (errors, dataVar) => {
+      throw new Error('123')
+    }
+  }, function (req, reply) {
+    reply.code(200).send(req.body.name)
+  })
+
+  fastify.inject({
+    method: 'POST',
+    payload: {
+      hello: 'michelangelo'
+    },
+    url: '/'
+  }, (err, res) => {
+    t.error(err)
+    t.deepEqual(res.json(), {
+      statusCode: 400,
+      error: 'Bad Request',
+      message: 'schema custom formatter error: 123'
+    })
+    t.strictEqual(res.statusCode, 400)
+    t.end()
+  })
+})
+
+test('test registering two instances with different errorFormatters', t => {
```

Comment:
Why this test?
I think it is the replica of `should call custom error formatter` but with 2 instances

Reply:
Your example clarified this for me, it should be the test that you described as far as I understand.

Reply:
@Eomm I have changed the `prefer route based error formatter over global one` accordingly, this test alone doesn't cover all the cases you asked for, but the other tests combined do cover the cases (except from the `fastify.register`), is this what you wanted, or did you want a single test that covers all of these cases?

Reply:
Unfortunately, the `register` is never called, so this is not covered.
Moreover I think that to let it works you will need to update the `pluginOverride.js` file